### PR TITLE
String enum converter

### PR DESCRIPTION
**Pull request by JeremyKuhne**
Adds public converter for converting enums to strings and vice-versa.

### CODE REVIEW

**Comment by JamesNK** [comment]
Json.NET supports EnumMemberAttribute to customize enum name values. It is a well used feature.

You don't need it in 3.0, but you will get a lot of requests for customizing enum names. Design for adding it in the future.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs [hunk 1]
```diff
@@ -2,168 +2,243 @@
...
                 {
                     ThrowHelper.ThrowJsonException();
                     return default;
                 }
 
+                // Try parsing case sensitive first
                 string enumString = reader.GetString();
-                if (!Enum.TryParse(enumString, out T value))
+                if (!Enum.TryParse(enumString, out T value)
+                    && !Enum.TryParse(enumString, ignoreCase: true, out value))
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: > Is this going to handle policies that do more than just change case?

It won't. We can look at more complicated parsing in the future if warranted. I'll make sure I'm super explicit in the comments.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs [hunk 2]
```diff
@@ -2,168 +2,243 @@
...
         {
-            if (TreatAsString)
+            // If strings are allowed, attempt to write it out as a string value
+            if (_converterOptions.HasFlag(EnumConverterOptions.AllowStrings))
             {
-                writer.WriteStringValue(value.ToString());
+                string output = value.ToString();
+                if (IsValidIdentifier(output))
+                {
+                    writer.WriteStringValue(_namingPolicy.ConvertName(output));
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: Ok. Caching when we're given a converter.

**Comment by JamesNK** [comment]
Check out https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Utilities/EnumUtils.cs

This is where Json.NET's enum heavy lifting logic lives. I basically build a contract of what the enum will look like for a given naming strategy.

**Comment by JeremyKuhne** [comment]
> You don't need it in 3.0, but you will get a lot of requests for customizing enum names. Design for adding it in the future.

It shouldn't be a problem to layer in that functionality post 3.0. Didn't want to make it too complicated this time given how late we're getting.

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+            // Numbers work by default
+            day = JsonSerializer.Parse<DayOfWeek>(@"4", options);
+            Assert.Equal(DayOfWeek.Thursday, day);
+
+            string json = JsonSerializer.ToString(DayOfWeek.Friday, options);
+            Assert.Equal(@"""Friday""", json);
+
+            // Try a unique naming policy
+            options = new JsonSerializerOptions();
+            options.Converters.Add(new JsonStringEnumConverter(new ToLower()));
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: When we support it, will do.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs [hunk 4]
```diff
@@ -2,168 +2,275 @@
...
+                }
             }
-            else if (s_enumTypeCode == TypeCode.UInt64)
+
+            if (!_converterOptions.HasFlag(EnumConverterOptions.AllowNumbers))
             {
-                // Use the ulong converter to prevent conversion into a signed\long value.
-                ulong ulongValue = Convert.ToUInt64(value);
-                writer.WriteNumberValue(ulongValue);
+                ThrowHelper.ThrowFormatException();
```
Comment (steveharter) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: This should be throwing JsonException unless you want a hard fail (end user bug). We no longer catch and re-throw `FormatException` as `JsonException`. Also applies to other cases here.

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using Xunit;
+
+namespace System.Text.Json.Serialization.Tests
```
Comment (steveharter) [inline src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: Please add a test that uses `[JsonConverter]` to specify the stringenum converter for a single property, but keeping the default converter as-is.

File: src/System.Text.Json/ref/System.Text.Json.cs [hunk 6]
```diff
@@ -440,6 +440,12 @@ public abstract partial class JsonConverter<T> : System.Text.Json.Serialization.
         public virtual void Write(System.Text.Json.Utf8JsonWriter writer, T value, System.Text.Json.JsonEncodedText propertyName, System.Text.Json.JsonSerializerOptions options) { }
         public abstract void Write(System.Text.Json.Utf8JsonWriter writer, T value, System.Text.Json.JsonSerializerOptions options);
     }
+    public sealed class JsonStringEnumConverter : System.Text.Json.Serialization.JsonConverterFactory
```
Comment (steveharter) [inline src/System.Text.Json/ref/System.Text.Json.cs]: Just for the record, the alternative here is to expose a single enum converter with a `bool` "TreatAsString" property. A new custom attribute deriving from `[JsonConverter]` would also need to be created in order to pass that `bool` on.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs [hunk 7]
```diff
@@ -2,168 +2,275 @@
...
 
         public override void Write(Utf8JsonWriter writer, T value, JsonEncodedText propertyName, JsonSerializerOptions options)
         {
-            if (TreatAsString)
+            // If strings are allowed, attempt to write it out as a string value
+            if (_converterOptions.HasFlag(EnumConverterOptions.AllowStrings))
             {
-                writer.WriteString(propertyName, value.ToString());
+                string original = value.ToString();
+                if (_nameCache != null && _nameCache.TryGetValue(original, out string transformed))
```
Comment (JamesNK) [inline src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: What is the lifetime of a converter? How long will this cache live for?

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+{
```
Comment (steveharter) [inline src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: Please add a test failure case that just uses `Enum` class as a property or return value from root (should fail).

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            Assert.Equal(@"""system, offline""", json);
+
+            // Undefined values should come out as a number (not a string)
+            json = JsonSerializer.ToString((FileAttributes)(-1), options);
+            Assert.Equal(@"-1", json);
+
+            // Not permitting integers should throw
+            options = new JsonSerializerOptions();
+            options.Converters.Add(new JsonStringEnumConverter(allowIntegerValues: false));
+            Assert.Throws<FormatException>(() => JsonSerializer.ToString((FileAttributes)(-1), options));
```
Comment (steveharter) [inline src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: JsonException?

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: Missed that change from yesterday.

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: Ok

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs (hunk 8 above)
Comment (JeremyKuhne) [reply src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: Not sure what you mean here. Use System.Enum directly?

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs (hunk 9 above)
Comment (JeremyKuhne) [reply src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: Yes, I didn't catch up with the exception changes from yesterday- fixing.

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs [hunk 10]
```diff
@@ -2,168 +2,275 @@
...
         {
-            TreatAsString = treatAsString;
+        }
+
+        public JsonConverterEnum(EnumConverterOptions options, JsonNamingPolicy namingPolicy)
+        {
+            _converterOptions = options;
+            if (namingPolicy != null)
+            {
+                _nameCache = new Dictionary<string, string>();
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: Isn't that the default?

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs (hunk 7 above)
Comment (JeremyKuhne) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: As long as JsonSerializerOptions it was attached to.

Comment (JamesNK) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: In that case you need to worry about thread-safety. Use ConcurrentDictionary

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: I thought it used string.Equals by default which uses the current culture.

[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: It uses `String.Equals(String, String)` which is ordinal (as is `String.Equals(String)`).

File: src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs (hunk 7 above)
Comment (JeremyKuhne) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: Ok.

Comment (JamesNK) [reply src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs]: Alternatively you could eagerly resolve the enum names at once and set to a dictionary. That would avoid locking every TryGetValue.

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: Apparently the converter attribute does not work with factories.

File: src/System.Text.Json/tests/Serialization/EnumConverterTests.cs (hunk 8 above)
Comment (steveharter) [reply src/System.Text.Json/tests/Serialization/EnumConverterTests.cs]: Yes something like `JsonSerializer.Parse<Enum>("1")` should fail gracefully (JsonException). I see we actually have a test for that -- so nevermind.

**Comment by JeremyKuhne** [comment]
Almost all of the failures were #38734. One failure was #37695.

**Pull request merged**